packages/xsnap/test/xsnap-eof.test.js

```diff
@@ -0,0 +1,298 @@
+  t.not(afterWaitError, undefined, 'vat.close() must produce an error');
+  t.is(vatExitSignal, null, 'vat exit signal must be null');
+  t.is(
+    vatExitCode,
+    ExitCode.E_IO_ERROR,
+    'vat exit code must indicate an IO error',
+  );
+  t.is(strOut, '', 'stdout must be empty');
+  t.is(strErr, expectedStderr, 'stderr must indicate expected error');
+}
+
+const testInterruption = test.macro(
+  /**
+   * @param {import('ava').ExecutionContext} t
+   * @param {(worker: ReflectiveWorker) => Promise<unknown>} beforeWait
+   * @param {(worker: ReflectiveWorker, message: Uint8Array) => Promise<Uint8Array>} onRequest
+   * @param {(worker: ReflectiveWorker) => Promise<unknown>} afterWait
+   * @param {(t: import('ava').ExecutionContext, results: Awaited<ReturnType<expectTermination>>) => void} verifyResults
+   */
+  async (t, beforeWait, onRequest, afterWait, verifyResults) => {
+    const handleCommand = message => {
+      // eslint-disable-next-line no-use-before-define
+      return onRequest(worker, message);
+    };
+    const worker = await spawnReflectiveWorker(handleCommand);
+    const results = await expectTermination(worker, beforeWait, afterWait);
+    verifyResults(t, results);
+  },
+);
+
+test(
+  'xsnap-worker complains while trying to READ an answer when pipes are closed',
+  testInterruption,
+  async worker => worker.vat.issueStringCommand('0'),
+  async function onRequest(worker, _message) {
+    // the worker is blocked on read here, so closing "toXsnap" pipe
+    // should cause an immediate read error on worker side.
+    worker.toXsnap.end();
+    worker.toXsnap.destroy();
+    return new Uint8Array();
+  },
+  worker => worker.vat.close(),
+  (t, results) =>
+    verifyStdError(t, results, 'Got EOF on netstring read. Has parent died?\n'),
+);
+
+test(
+  'xsnap-worker complains while trying to WRITE an answer when pipes are closed',
+  testInterruption,
+  async worker => worker.vat.issueStringCommand('0'),
+  async function onRequest(worker, _message) {
+    // The worker is blocked on read here, so closing "fromXsnap" pipe
+    // does not cause an immediate exit. However, an attempt to send an
+    // answer back to us will cause a write error with a SIGPIPE.
+    worker.fromXsnap.end();
+    worker.fromXsnap.destroy();
+    return new Uint8Array();
+  },
+  worker => worker.vat.close(),
+  (t, results) =>
+    verifyStdError(t, results, 'Caught SIGPIPE. Has parent died?\n'),
+);
+
+test(
+  'xsnap-worker exits quietly when pipes are closed in quiescent state',
+  testInterruption,
+  async function beforeWait(worker) {
+    // Simulating an orderly shutdown in quiescent state by closing both pipes.
+    // The worker is blocked on read here, so we should close "fromXsnap" pipe first
+    worker.fromXsnap.end();
+    worker.fromXsnap.destroy();
+    worker.toXsnap.end();
+    worker.toXsnap.destroy();
+  },
+  async function onRequest(_worker, _message) {
+    return new Uint8Array();
```

Comment:
My understanding is that we should never see a request/query here, so throwing instead would be more appropriate.